flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/io/checkpointing/AlternatingController.java

```diff
@@ -188,12 +266,12 @@ private boolean canTimeout(CheckpointBarrier barrier) {
         return barrier.getCheckpointOptions().isTimeoutable()
                 && barrier.getId() <= lastSeenBarrier
                 && barrier.getCheckpointOptions().getAlignmentTimeout() * 1_000_000
-                        < (System.nanoTime() - firstBarrierArrivalTime);
+                        < (clock.relativeTimeNanos() - lastBarrierArrivalTime);
     }
 
-    private long getArrivalTime(CheckpointBarrier announcedBarrier) {
-        return announcedBarrier.getCheckpointOptions().isTimeoutable()
-                ? System.nanoTime()
-                : Long.MAX_VALUE;
+    /** A provider for a method to register a delayed action. */
+    @FunctionalInterface
+    public interface DelayedActionRegistration {
+        void schedule(Callable<?> callable, Duration delay);
```

Comment:
I'm thinking about cancellation of the scheduled timers. Depending on the effective checkpoint interval and alignmentTimeout there can be up to alignmentTimeout / interval active timers.

E.g. if interval = 1s and timeout = 20m then numTimers = 1200, firing every second. This is further multiplied by the number of tasks for a single TM.

Or am I missing something?

Reply:
First of all I think the alignment timeout plays no role in the number of timers. There will always be at most two timers for each aligned checkpoints (one for the announcement and one for the first barrier). I agree it is multiplied by the number of tasks for a single TM.

We briefly discussed that with @AHeise but we thought the number of timers will be insignificant compared to the number of records that we can process. Especially as in that case the executed code for a timer will be a single boolean condition evaluation, as such number of timers can happen only in case of successful aligned checkpoints.

Reply:
> First of all I think the alignment timeout plays no role in the number of timers

Could you explain why? My reasoning is:
- each checkpoint causes timer registration
- timer is not unregistered in the current implementation
- timer fires after alignment timeout
So during the alignment timeout, pedning timers accumulate.

Reply:
Yes, they do accumulate, but it does not affect the number of timers. As you said each checkpoint causes timer registration. With a single checkpoint every second you will have a single(two) timer(s) firing every second (with a delay of the timeout alignment).

Reply:
Unless what you meant is not that 1200 timers will fire every second, but that we will have that many timers registered.

Reply:
Yes, I meant 1200 **pending** timers, each one firing every second.

Reply:
I am cancelling outdated timers in the new version. PTAL.